Update 07a-troubleshooting-wsl-update.md

amesserla opened the pull request:
Add description of IP configuration for Windows and WSL

amesserla committed:
Update 07a-troubleshooting-wsl-update.md

Add description of IP configuration for Windows and WSL

aregtech commented on docs/wiki/07a-troubleshooting-wsl-update.md:
```diff
@@ -117,3 +119,49 @@ If the issue persists, reset the WSL network configuration:
+
+---
+
+## Issue 3: Configuring communication between Windows and WSL
```
Could you please include a detailed description of **Issue 3** before presenting the solution? It’s important for users to clearly understand the problem first. For example, you might explain it as:  
 
*"There is no communication between applications running on WSL and the system-managed services, such as `mcrouter` and/or `logcollector`, running on Windows."*  

This context will help users better grasp the issue before moving on to the resolution.

aregtech commented on docs/wiki/07a-troubleshooting-wsl-update.md:
```diff
@@ -117,3 +119,49 @@ If the issue persists, reset the WSL network configuration:
+
+>[!IMPORTANT]
+>This assumes WSL is using the default networking mode, NAT. Changing this setting to Mirror may allow ``localhost`` to work as intended, but this has not been tested.]
+
```
If there is public available documentation explaining changing **setting to Mirror**, would be fine include link. Not everyone may understand what is the meaning *changing setting to Mirror*.

aregtech commented on docs/wiki/07a-troubleshooting-wsl-update.md:
```diff
@@ -117,3 +119,49 @@ If the issue persists, reset the WSL network configuration:
+>This assumes WSL is using the default networking mode, NAT. Changing this setting to Mirror may allow ``localhost`` to work as intended, but this has not been tested.]
+
+> [!NOTE]
+> When connecting a client on Windows to a server or router on WSL, the default ``localhost`` settings will allow the connection.
```
This is an excellent example of explaining the issue: services running on WSL with `localhost` settings are accessible, whereas services running on Windows with `localhost` settings are not. I suggest moving this explanation to the top of the document, where the issue is initially described, to provide clear context to the users.

aregtech commented on docs/wiki/07a-troubleshooting-wsl-update.md:
````diff
@@ -117,3 +119,49 @@ If the issue persists, reset the WSL network configuration:
+      logger::*::address::tcpip   = NEW_IP
+   ```
+
+   changed so that ``NEW_IP`` is replaced with the IP address found above.
````
It might be clearer to replace `NEW_IP` with `<wsl-ip-address>` or `<windows-ip-address>`. This would make it easier for users to understand exactly where and which IP address they need to specify.

aregtech commented on docs/wiki/07a-troubleshooting-wsl-update.md:
````diff
@@ -117,3 +119,49 @@ If the issue persists, reset the WSL network configuration:
+       logger::*::address::tcpip   = 0.0.0.0
+   ```
+
+   with ``0.0.0.0`` or an IP address that is configured to accept external connections replacing the defualt ``localhost``.
````
In my case, I conducted the tests as follows:

1. **Windows**: Run `mcrouter` on Windows.  
2. **WSL**: Run `logcollector` on WSL.  
3. Use `ipconfig` on Windows and `ifconfig` in the WSL terminal to obtain the IP addresses of the PC running Windows and WSL.  
4. Modify the `areg.init` file on both Windows and WSL to specify the following:
   - `router::*::address::tcpip = <windows-ip-address>`  
   - `logger::*::address::tcpip = <wsl-ip-address>`  
5. Start `mcrouter` on Windows and `logcollector` on WSL.  
6. Launch `logobserver` and start AREG-based applications on either Windows, WSL, or a mix of both.

**Result**: The applications started successfully, and communication worked seamlessly.  

Could you please also specify this as **Solution 2**?

aregtech commented on docs/wiki/07a-troubleshooting-wsl-update.md:
````diff
@@ -117,3 +119,49 @@ If the issue persists, reset the WSL network configuration:
+      ip route show | grep -i default | awk '{print $3}'
+   ```
+
+2. In the WSL configuration file, the address for any services running as a router or server must be changed to the IP address found above. For example, to connect to an instance of ``mcrouter`` and ``logcollector`` running on Windows, the configuration file should have the lines
````
> *the configuration file*

Could you please specify the exact file name, `areg.init` (probably with the link to the file in repo), so that users clearly know which file needs to be modified? This will help avoid any confusion and ensure accurate updates.

aregtech reviewed (changes requested):
@amesserla,  
Thank you very much for updating the document.  
Overall, it looks good. However, a few adjustments could make the document clearer and more helpful.  
Kindly review my notes for further improvement.

amesserla committed:
Update 07a-troubleshooting-wsl-update.md

Add description of problem, moved Note, gave resource for Mirrored setting

amesserla committed:
Update 07a-troubleshooting-wsl-update.md

``areg.init`` file specified and linked

amesserla committed:
Update 07a-troubleshooting-wsl-update.md

Add suggested Solution  2.

amesserla committed:
Update 07a-troubleshooting-wsl-update.md

Update Table of Contents

amesserla committed:
Update 07a-troubleshooting-wsl-update.md

Change ``NEW_IP`` to ``<windows-ip-address>`` for specificity and moved Tip with reference to WSL network access to end of issue.

amesserla committed:
Update 07a-troubleshooting-wsl-update.md

Edit for consistent formatting

amesserla commented:
I have made the requested changes. Let me know if there are any more improvements you think I could make!

aregtech commented:
Hi @amesserla,  

Thank you for your contributions! The updates look much better now!  

I’ve reviewed the file and made some updates to further enhance clarity and accuracy. Since I cannot directly push changes to the pull request you’ve created, I’ve created a new branch with the modified file for your review: Modified [07a-troubleshooting-wsl-update](https://github.com/aregtech/areg-sdk/blob/pull/480/docs/wiki/07a-troubleshooting-wsl-update.md) file.  

Key Updates:  
1. Corrected spelling errors.  
2. Updated links to the `areg.init` file.  
3. Refined some wording for better clarity. Specifically, I noticed you mentioned "services and servers" in a few places. In the context of the `areg-sdk`, **services** often refer to **physical** servers. To avoid confusion, I clarified that the AREG Framework supports both **logical** and **physical** servers, enabling software components to act as **logical servers**, **logical clients**, or both. However, this distinction isn’t directly relevant to the physical IP connection discussed here, so I adjusted the phrasing accordingly.  

Could you please review the changes, especially in *Solution 1*, to confirm if everything is accurate? If you agree with the updates, feel free to copy and push them to the pull request.  

Many thanks in advance for your time and efforts!

aregtech reviewed (changes requested):
Please compare with the changes in https://github.com/aregtech/areg-sdk/blob/pull/480/docs/wiki/07a-troubleshooting-wsl-update.md

amesserla committed:
Update 07a-troubleshooting-wsl-update.md

Remove repeated Solution 2:

amesserla commented:
Hi @aregtech, 

I'm away from the computer I test on for a few days, so I can't answer this myself right now. In the re-writing of Solution 2: it appears you don't need to edit the ``router::*::address::tcpip = localhost`` line in ``areg.init`` on Windows. Wouldn't this cause ``mcrouter`` to listen on ``127.0.0.1`` and fail to connect with ``logcollector`` on WSL?

Thank you for clarifying the differences between the usage of services and servers. I'll look carefully at your usage so I can note the differences when looking through other parts of the documentation.

aregtech commented:
Hi @amesserla,  

No problem at all, there's no rush! We can continue once you're back.  

> Wouldn't this cause mcrouter to listen on 127.0.0.1 and fail to connect with logcollector on WSL?  

That's a good question, and to be honest, I have no clear answer. It seems testing of various configurations is needed. It looks like you have a deeper experience and understanding of WSL and Windows networking than I do. If you could find the time to test multiple variations and document your observations, it would be incredibly helpful. Your insights would definitely contribute to improving this aspect of the project.

amesserla commented:
Hi @aregtech,

I am back to my pc and did some testing, and here is what I found. 

The setup had `mcrouter` running on Windows, and `logcollector` on WSL, with `logobserver` running one instance on Windows and one instance on WSL. I tested all combinations of the following 4 iterations of `areg.init`.

Config 1 contained:
```
router::*::address::tcpip  = localhost
logger::*::address::tcpip = <wsl-ip>
```

Config 2 contained:
```
router::*::address::tcpip  = <windows-ip>
logger::*::address::tcpip = <wsl-ip>
```

Config 3 contained:
```
router::*::address::tcpip  = localhost
logger::*::address::tcpip = localhost
```

 Config 4 contained:
```
router::*::address::tcpip  = <windows-ip>
logger::*::address::tcpip = localhost
```

Results:

The results are displayed in the following table where the number on top specifies the modification to `areg.init` used on Windows and the left column specifies the modification to `areg.init` used on WSL. Specifically, the 4th row and 1st column represents how I understood the suggested configuration from [https://github.com/aregtech/areg-sdk/blob/pull/480/docs/wiki/07a-troubleshooting-wsl-update.md](https://github.com/aregtech/areg-sdk/blob/pull/480/docs/wiki/07a-troubleshooting-wsl-update.md)

|  | 1 | 2 | 3 | 4 |
|---|---|---|---|---|
| 1 | ✔️ | ✔️ | ❌ | ❌ |
| 2 | ✔️ | ✔️ | ❌ | ❌ |
| 3 | ❌ | ❌ | ✔️ | ✔️ |
| 4 | ❌ | ❌ | ✔️ | ✔️ | 

✔️ - indicates that both instances of `logobserver` showed a connection to `mcrouter` and `logcollector` showed a connection to `mcrouter` and both instances of `logobserver`
❌ -  indicates that neither instance of `logobserver` showed a connection to `mcrouter` and `logcollector` only showed a connection to the instance of `logobserver` running on WSL

In particular, for a successful connection, the line `logger::*::address::tcpip` must match between the copies of `areg.init` on WSL and on Windows, but the line `router::*::address::tcpip` does not need to match. 

I looked back at the original issue of running both `mcrouter` and `logcollector` on Windows, and `logobserver` on WSL and found that the connection works if the line `logger::*::address::tcpip = localhost` is changed to `<windows-ip>` in both `areg.init` files, but changing the line `router::*::address::tcpip  = localhost` is not necessary in either. This same configuration works with `mcrouter` running on WSL and `logobserver` running on either.

-----------

My guess as to what is happening is that the `logcollector` application is functioning as the server that `mcrouter` and `logobserver` are connecting to. If `logcollector` is running on WSL, it can either be accessed by `localhost` (since Windows has access to the WSL `localhost`) or by `<wsl-ip>`, but which ip address is used must be specified to the Windows applications via `areg.init`. If `logcollector` is running on Windows, the ip address for `logcollector` must be changed from `localhost` (because WSL does not have access to the Windows `localhost`). 

I should mention that I have only done testing with `mcrouter`, `logcollector`, and `logobserver`. It may be that sending or receiving information between another application and `mcrouter` (if this is something that can happen) would require other configuration changes. Do you have a recommended way of testing this?

aregtech commented:
Hi @amesserla .
Happy New Year! :)

Thank you very much for your detailed report—this is truly outstanding work.

About applications communicating via `mcrouter`, I believe it works similarly to communication with `logcollector`. If I understand correctly, the best solution would be to set both `<windows-ip>` and `<wsl-ip>` in the configuration. We could present this as the recommended approach while noting that using `localhost` in the configuration for applications on WSL can serve as an alternative solution.

Could you please incorporate a summary of this into the document? Also, since your report is very detailed and helpful, it would be great to include a permanent link to [your message](https://github.com/aregtech/areg-sdk/pull/480#issuecomment-2568508064). If you're unsure whether GitHub keeps pull requests long-term, you could copy your message to [this ticket](https://github.com/aregtech/areg-sdk/issues/479) and refer to it in the document. Alternatively, you could include the message directly in the document.  

Please feel free to choose the approach you find most appropriate. The goal is to give users clear and detailed information.  

Thanks again for your excellent work!

aregtech commented:
> ✔️ - indicates that both instances of `logobserver` showed a connection to `mcrouter` and `logcollector` showed a connection to `mcrouter` and both instances of `logobserver`
> ❌ - indicates that neither instance of `logobserver` showed a connection to `mcrouter` and `logcollector` only showed a connection to the instance of `logobserver` running on WSL

Hi @amesserla,  

I wanted to quickly clarify to ensure there’s no misunderstanding:  

1. **`logcollector`** serves as the central server for all applications that are registered as "log sources" to send log messages.  
2. **`logobserver`** is a client of `logcollector` and does not have a direct physical connection to `mcrouter`.  
3. **`mcrouter`** acts as a physical server for all applications but also functions as a client of `logcollector` to send log messages.

amesserla commented:
Hi @aregtech 

> I wanted to quickly clarify to ensure there’s no misunderstanding:

This helped clarify the different applications for me a lot. 

> About applications communicating via `mcrouter`, I believe it works similarly to communication with `logcollector`

I did some more testing to confirm this, using the [`20_pubdatarate`](https://github.com/aregtech/areg-sdk/tree/master/examples/20_pubdatarate) example to confirm that if applications running on WSL want to communicate with `mcrouter` running on Windows, then both `areg.init` files (on WSL and Windows) must be updated to specify `router::*::address::tcpip  = <windows-ip>`. This fits into my understanding that if an application on WSL wants to make a connection with a server on Windows, acceptable ip addresses must be specified.

Specifically:

I launched `mcrouter` and `20_pubclient` on Windows and `20_pubserver` on WSL and started `20_pubserver`. When the default `router::*::address::tcpip  = localhost` was left unchanged in either copy of `areg.init`, no connection occurred. However, after changing  `router::*::address::tcpip  = <windows-ip>` in both `areg.init` files, the connection occurred and data was exchanged as intended.

With `mcrouter` running on WSL and `20_pubclient` and `20_pubserver` on Windows , using `localhost` in both `areg.init` files, or using `<wsl-ip>` in both files, worked as intended.

> the best solution would be to set both `<windows-ip>` and `<wsl-ip>` in the configuration.

I agree that this is the best solution. If it sounds good to you, I will rewrite the documentation to offer this as solution 1, and use my explanations above to offer a solution 2 which includes the fact you can use `localhost` for applications running as a server on WSL, even if clients running on Windows want to connect to that server, but with the caveat that the specified ip addresses must match between the two copies of `areg.init`.

aregtech commented:
Hi @amesserla 

> If it sounds good to you, I will rewrite the documentation

Yes please, this sounds absolutely fine.

amesserla committed:
Update 07a-troubleshooting-wsl-update.md

Redid solutions to Issue 3

aregtech committed:
Update 07a-troubleshooting-wsl-update.md

Replaced W**I**ndows with W**i**ndows.
Added a link to the **Solution 2** reference
Very slight formatting by removing spaces.

aregtech reviewed (approved):
@amesserla , thank you very much. It is informative and good described.
